### PR TITLE
 server: use leader lease to determine tso service validity

### LINKED ISSUE #1661: Use monotonic clock to ensure TSO safty
During leader election, the TSO could be corrupted if the notify message of keep alive channel is not received in time. On this situation, old leader and new leader can serve timestamps at the same time.
Possible fix: Store lease expire time (in monotonic_raw fashion) and check lease for every received GetTS requests.

### PR DESCRIPTION
**Pull request by disksing**
STATUS: It seems to be better merged after https://github.com/pingcap/pd/pull/1668. You can review it now and give comments.

### What problem does this PR solve? 
Currently, the safety of timestamp allocation relies on the leader flag be updated in time when lease timeout.
There is a small possibility (for instance, process paused, runtime schedule slow) that the watch channel does not notify in time, which may cause 2 PD servers serve timestamp at the same time. Then the transaction may become corrupted.

Another issue is when a PD server becomes leader, it won't serve timestamps until the leader flag is set. But the leader flag is set after loading all regions from storage, which may take considerably long time for a large cluster.

This PR fixes #1661 and part of #1658 

### What is changed and how it works?

Introduce `LeaderLease` to periodically renew lease and update expire time.

Timestamp service only relies on the lease expire time.

Note that if 2 `time.Time`s both have monotonic part, comparison of them will be evaluated by monotonic time, which is not affected by wall clock changes.

### Check List 

Tests 
 - Unit test

Side effects
 - Increased code complexity
 - May introduce bugs

Related changes
 - Need to be included in the release notes

### CODE REVIEW

File: server/lease.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+// IsExpired checks if the lease is expired. If it returns true, current PD
+// server should step down and try to re-elect again.
+func (l *LeaderLease) IsExpired() bool {
+	return time.Now().After(l.expireTime.Load().(time.Time))
+}
+
+// KeepAlive auto renews the lease and update expireTime.
+func (l *LeaderLease) KeepAlive(ctx context.Context) {
```
Comment (shafreeck) [inline server/lease.go]: I found that the main enhancement to maintain the `expireTime`, so why not use etcd's `KeepAlive` and update `expireTime` when got a response from the keepalive channel?

File: server/lease.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	timeCh := l.keepAliveWorker(ctx, l.leaseTimeout/3)
```
Comment (rleungx) [inline server/lease.go]: Why use `3` here?

File: server/lease.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	l.ID = leaseResp.ID
+	l.leaseTimeout = time.Duration(leaseTimeout) * time.Second
+	l.expireTime.Store(start.Add(time.Duration(leaseResp.TTL) * time.Second))
+	return nil
+}
+
+// Close releases the lease.
+func (l *LeaderLease) Close() error {
+	return l.lease.Close()
```
Comment (nolouch) [inline server/lease.go]: May we can revoke the `lease` before close?

File: server/lease.go (hunk 1 above)
Comment (Luffbee) [reply server/lease.go]: > I found that the main enhancement is to maintain the `expireTime`, so why not use etcd's `KeepAlive` and update `expireTime` when got a response from the keepalive channel?

I think the problem here is that `KeepAlive` only returns TTL. We cannot know the "start time" of the TTL (`time.Now()` is not the answer), so we cannot calculate the `expireTime`.

Comment (shafreeck) [reply server/lease.go]: OK, thanks

File: server/leader.go [hunk 4]
```diff
@@ -36,6 +35,8 @@ import (
 // The timeout to wait transfer etcd leader to complete.
 const moveLeaderTimeout = 5 * time.Second
 
+const leaderTickInterval = 50 * time.Millisecond
```
Comment (Luffbee) [inline server/leader.go]: Maybe add some comments about the relation of `leaderTickInterval` and `tso.UpdateTimestampStep`?

File: server/lease.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+// Grant uses `lease.Grant` to initialize the lease and expireTime.
+func (l *LeaderLease) Grant(leaseTimeout int64) error {
+	start := time.Now()
+	ctx, cancel := context.WithTimeout(l.client.Ctx(), requestTimeout)
+	leaseResp, err := l.lease.Grant(ctx, leaseTimeout)
+	cancel()
+	if cost := time.Since(start); cost > slowRequestTime {
+		log.Warn("lease grants too slow", zap.Duration("cost", cost))
+	}
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (disksing) [reply server/lease.go]: Good idea.

File: server/lease.go (hunk 2 above)
Comment (disksing) [reply server/lease.go]: Arbitrary value borrowed from etcd's keep alive code.

File: server/leader.go (hunk 4 above)
Comment (disksing) [reply server/leader.go]: Actually no special relationship, they can run independently.

Comment (Luffbee) [reply server/leader.go]: `leaderTickInterval` controls the frequency of calling `UpdateTimestamp`. And in `tso` package, it assumes the frequency is controlled by `tso.UpdateTimestampStep`.

Comment (disksing) [reply server/leader.go]: Updated. Use different ticker now 😄

**Comment by disksing** [comment]
Please take another look. @rleungx @nolouch

**Comment by disksing** [comment]
Created an issue for the test failure: https://github.com/pingcap/pd/issues/1693

**Comment by nolouch** [comment]
@disksing should we cherry pick to 3.0,3.1?

**Comment by disksing** [comment]
@nolouch I think no.

**Pull request merged**